mlflow/tensorflow/__init__.py

```diff
@@ -888,6 +996,99 @@ def _log_early_stop_callback_metrics(callback, history, metrics_logger):
         if metric_key is not None:
             metrics_logger.record_metrics(restored_metrics, stopped_epoch + 1)
 
+    def _log_keras_model(history, args):
+        def _get_input_data_slice():
+            """
+            Generates a sample ndarray or dict (str -> ndarray)
+            from the input type 'x' for keras ``fit`` or ``fit_generator``
+
+            :return: a slice of type ndarray or
+                     dict (str -> ndarray) limited to ``INPUT_EXAMPLE_SAMPLE_ROWS``
+            """
+            input_training_data = args[0]
+            input_example_slice = None
+            if isinstance(input_training_data, tensorflow.keras.utils.Sequence):
+                input_training_data = input_training_data[:][0]
+
+            if isinstance(input_training_data, np.ndarray):
+                input_example_slice = input_training_data[:INPUT_EXAMPLE_SAMPLE_ROWS]
+            elif isinstance(input_training_data, tensorflow.data.Dataset):
+                steps = 1
+                if history.params is not None and "steps" in history.params:
+                    steps = history.params["steps"]
+
+                def _extract_n_steps(input_example_n_steps):
+                    # Example:
+                    # when the tensorflow.data.Dataset is batched
+                    # (Total number of steps (batches of samples)),
+                    # extract first ``INPUT_EXAMPLE_SAMPLE_ROWS`` samples from the first batch
+                    if steps > 1:
+                        return np.array(
+                            [
+                                v[0][:INPUT_EXAMPLE_SAMPLE_ROWS]
+                                for v in input_example_n_steps.take(1).as_numpy_iterator()
+                            ]
+                        )[0]
+                    return np.array(
+                        [
+                            v[0]
+                            for v in input_example_n_steps.take(
+                                INPUT_EXAMPLE_SAMPLE_ROWS
+                            ).as_numpy_iterator()
+                        ]
+                    )
+
+                input_example_slice = _extract_n_steps(input_training_data)
+            elif isinstance(input_training_data, dict):
+                input_example_slice = {
+                    k: np.take(v, range(0, INPUT_EXAMPLE_SAMPLE_ROWS))
+                    for k, v in input_training_data.items()
+                }
+            else:
+                warnings.warn(
+                    "Tensorflow keras autologging only "
+                    "supports input types of: numpy.ndarray, "
+                    "dict(<key> -> numpy.ndarray), tensorflow.data.Dataset, "
+                    "or tensorflow.keras.utils.Sequence"
```

Comment:
Can we adjust this to make it clearer that* input example logging* is only supported for these types? The rest of the autologging functionality still works.